[solver/pds]

joanrue opened the pull request:
Add Primal Dual Splitting family of solvers including: Condat-Vu (PDS), ChambollePock (CPS), DouglasRachford (DRS), Forward-Backward (FBS) and Primal-Dual three Operator (PD3O) splitting.

joanrue committed:
[solver/PDS] Add PDS associated subproblems' algorithms

joanrue committed:
[operator/ProxFunc/__mul__][BugFix] avoid squeezing ProxLike maps
ProxLike maps (ProxFunc, ProxDiffFunc) cannot be squeezed, and they are already FuncLike.

joanrue committed:
[runtime/coerce][BuxFix] Allow None inputs

joanrue committed:
[solver/pds][EXAMPLE] plot observed data

joanrue committed:
[solver/pds/example][BugFix] First derivative had wrong default axis

joanrue committed:
[operator/LinOp/lipschitz][BugFix] return float instead of np.array (cupy compatibility)

joanrue committed:
[_dev/L1Norm/prox][BugFix] mod. agnostic abs() and clip() funcs

joanrue committed:
[_dev/FirstDerivative][BugFix] First derivative had wrong default axis

joanrue committed:
[operator/asarray][BugFix] pass arr.mod. as input for cupy compatibility

joanrue committed:
[solver/pds][DOC] add docstring to PDS algo (reused from v1)

joanrue committed:
[solver/pds][refactor] change x_init/z_init to x0/z0

joanrue committed:
[solver/pds][clean] remove TODOs

joanrue committed:
[solver/pds][refactor] change x_init/z_init to x0/z0

joanrue committed:
[solver/pds][ENH] add short-names for family of PDS algorithms

joanrue committed:
[solver/pds][clear] remove test example

joanrue committed:
[solver/pds][DOC] add doc for pds subproblems

joanrue committed:
[solver/pds][BufFix] fix input parameter & defaults for pds subproblems

joanrue committed:
[solver/pds][ENH] add PD3O algorithm

joanrue committed:
[doc/solver] add pds module to doc

matthieumeo commented on src/pycsou/abc/operator.py:
```diff
@@ -1432,7 +1438,8 @@ def __mul__(self: "ProxFunc", other: MapLike) -> MapLike:
             )
-        return f.squeeze()
+
+        return f.squeeze() if f is NonProxLike else f
```
Can you remind me why we need to handle the ``NonProxLike`` case separately?

matthieumeo commented on src/pycsou/abc/operator.py:
```diff
@@ -1783,7 +1790,7 @@ def lipschitz(self, recompute: bool = False, algo: str = "svds", gpu: bool = Fal
         if recompute or (self._lipschitz == np.infty):
             kwargs.update(dict(k=1, which="LM", gpu=gpu))
-            self._lipschitz = self.svdvals(**kwargs)
+            self._lipschitz = self.svdvals(**kwargs).item()
```
Why the ``item()``? If this change is really needed maybe it should be done on the output of the ``svdvals`` method directly?

matthieumeo commented on src/pycsou/opt/solver/pds.py:
```diff
@@ -0,0 +1,965 @@
+
+    Notes
+    -----
+    The *Primal Dual Splitting (PDS)* method is described in [PDS]_ (this particular implementation is based on the pseudo-code Algorithm 7.1 provided in [FuncSphere]_ Chapter 7, Section1).
```
You should add the references somewhere for them to be operational. See how this was done in pycsou v1.

matthieumeo commented on src/pycsou/opt/solver/pds.py:
```diff
@@ -0,0 +1,965 @@
+        (..., N) initial point(s) for the dual variable.
+        If None (default), then use x0 as the initial point(s) for the dual variable as well.
+    tau: Real
+        Primal step size
```
Missing final dot.

matthieumeo commented on src/pycsou/opt/solver/pds.py:
```diff
@@ -0,0 +1,965 @@
+        Dual step size
+    rho: Real
+        Momentum parameter
+    beta: Real
```
Beta should be provided in ``__init__`` (it's fixed for a given ``f``, we cannot change it so no reason to have it in ``fit()``.

matthieumeo commented on src/pycsou/opt/solver/pds.py:
```diff
@@ -0,0 +1,965 @@
+        f: typ.Optional[pyco.DiffFunc] = None,
+        g: typ.Optional[pyco.ProxFunc] = None,
+        h: typ.Optional[pyco.ProxFunc] = None,
+        K: typ.Optional[pyco.LinOp] = None,
```
``K`` should be allowed to be of more general type: ``pyco.DiffMap``. Tbd together.

matthieumeo commented on src/pycsou/opt/solver/pds.py:
```diff
@@ -0,0 +1,965 @@
+        if h is not None:
+            if K is None:
+                self._K = pyclo.IdentityOp()
+                self._K._lipschitz = 1
```
No need the Lipschitz constant of the ``IdentityOp`` is set to one at construction time.

matthieumeo commented on src/pycsou/opt/solver/pds.py:
```diff
@@ -0,0 +1,965 @@
+                self._K = pyclo.IdentityOp()
+                self._K._lipschitz = 1
+            else:
+                if (K.shape[0] == h.shape[1]) or (h.shape[1] is None):
```
We don't check the shapes of ``f, g,h`` nor their consistency so I guess we shouldn't check the consistency of the shapes of ``K`` and ``h`` neither. 

I would rewrite lines 159-173 as:

```python3

if h is not None: 
   self._K =  pyclo.IdentityOp() if (K is None) else K
else:
   if K is None:
      self._K = pyclo.NullOp()
   else: 
      raise ValueError('Optional argument ``h`` mut be specified if ``K`` is not None.')

matthieumeo commented on src/pycsou/opt/solver/pds.py:
```diff
@@ -0,0 +1,965 @@
+    :py:class:`~pycsou.opt.solver.pds.PDS`, :py:class:`~pycsou.opt.solver.pds.ChambollePockSplitting`, :py:class:`~pycsou.opt.solver.pds.DouglasRachford`
+    """
+
+    def __init__(
```
You should add the argument ``beta`` and store its value in instance attribute ``_beta``.

matthieumeo commented on src/pycsou/opt/solver/pds.py:
```diff
@@ -0,0 +1,965 @@
+    def set_step_sizes(self, tau: typ.Optional[pyct.Real], sigma: typ.Optional[pyct.Real]) -> typ.Tuple[float, float]:
+        r"""
+        Set the x/z step sizes.
+        Returns
```
you should always have blank line after the 1st line of docstrings (this way Sphinx interprets the first line as the short description of the function).

matthieumeo replied:
More generally always put blank lines after/before functions or the formatting will be wrong.

matthieumeo commented on src/pycsou/opt/solver/pds.py:
```diff
@@ -0,0 +1,965 @@
+        mst = self._mstate  # shorthand
+        mst["x"] = mst["x_prev"] = x0
+        mst["z"] = mst["z_prev"] = self.set_dual_variable(z0)
+        mst["beta"] = self.set_beta(beta)
```
SHould go in __init__.

matthieumeo commented on src/pycsou/opt/solver/pds.py:
```diff
@@ -0,0 +1,965 @@
+           \tau=\sigma=\Vert\mathbf{K\Vert_{2}^{-1}.}
+        """
+
+        if (tau is not None) and (sigma is None):
```
The first two ``if`` cases are wrong: one cannot just set ``tau=sigma`` for any value of ``tau/sigma``. We should discuss that.

matthieumeo commented on src/pycsou/opt/solver/pds.py:
```diff
@@ -0,0 +1,965 @@
+            log_var=log_var,
+        )
+
+    @pycrt.enforce_precision(i=["x0", "z0", "tau", "sigma", "rho"], allow_None=True)
```
You have to set ``tau`` and ``sigma`` optimally here too.

matthieumeo replied:
You can reuse ``set_step_sizes`` from the PDS class with ``beta=0``.

matthieumeo commented on src/pycsou/opt/solver/pds.py:
```diff
@@ -0,0 +1,965 @@
+        )
+
+    @pycrt.enforce_precision(i=["x0", "z0", "tau", "sigma", "rho"], allow_None=True)
+    def m_init(
```
Again we must set ``tau`` here in case the user does not provide it and make sure the algorithm converges.

matthieumeo commented on src/pycsou/opt/solver/pds.py:
```diff
@@ -0,0 +1,965 @@
+        mst = self._mstate  # shorthand
+        mst["x"] = mst["x_prev"] = x0
+        mst["z"] = mst["z_prev"] = self.set_dual_variable(z0)
+        mst["tau"] = tau
```
Again what happens if ``tau`` is not provided?

matthieumeo commented on src/pycsou/opt/solver/pds.py:
```diff
@@ -0,0 +1,965 @@
+        verbosity: int = 1,
+        log_var: pyct.VarName = ("x",),
+    ):
+        super().__init__(
```
This ``__init__`` is common to the ``PDS`` class. We should define a super class where this is done and have both classes inherit from that class.

matthieumeo commented on src/pycsou/opt/solver/pds.py:
```diff
@@ -0,0 +1,965 @@
+
+    **Remark 2:**
+
+    Assume that the following holds:
```
PD3O has different convergence properties and step size rules!

matthieumeo commented on src/pycsou/opt/solver/pds.py:
```diff
@@ -0,0 +1,965 @@
+FBS = ForwardBackwardSplitting
+
+
+class PD3O(pycs.Solver):
```
This class is extremely redundant with PDS. We need a superclass. Also the step sizes are incorrectly set here since PD3O has different convergence properties.

matthieumeo reviewed (changes requested):
Nice work Joan thanks ! There is a few things we need to correct, especially regarding the automatic determination of step sizes which is currently partially wrong or incomplete. Also PD3o and PDS are too redundant we need to have a super class. FInally PDS algorithms also work for DiffMap Ks so we should generalise to this case too (only when we are done fixing the rest though).

joanrue committed:
[pre-commit/black][BugFix] update black version
Related to this known error

joanrue committed:
[LinOp/svdvals] return value, not array
Otherwise this can modify the shape of parameters when using the Lipschitz for automatic parameter tuning

joanrue committed:
[References] add PD3O reference, and rename PDS to CVS
Now PDS is used for the as the abstract class of solver including CondatVuSplitting (CVS) and PD3O

joanrue committed:
[solver/pds][refactoring] PDS as abstract base class for CVS and PD3O
The convergence conditions have been corrected for both CVS and PD3O. Allowing DiffMap instead of LinOp is left for future (marked in the code as #TODO).

joanrue committed:
[solver/pds][clean] remove unused/commented code

joanrue committed:
[solver/pds][DOC] clean, update to v2 format and fix typos

joanrue committed:
[solver/pds/pd3o] fix convergence criteria

matthieumeo committed:
[PDS] Changes to the API and small bug fixes.

joanrue committed:
[references] add refs to papers for PD3O convergence conditions

joanrue committed:
[abc/operator][BugFix] svdvals output type hint float instead of NDArray

joanrue committed:
[solver/pds] changes to the API and typos

joanrue committed:
[solver/pds][DOC] convergence conditions on parameters (CV/PD3O)

joanrue committed:
[solver/pds/PD3O][BugFix] fix set_step_sizes() bugs

joanrue committed:
[solver/pds] set_momentum_term() moved to specific cases (CV/PD3O)"

matthieumeo committed:
[PDS] Changes to the API and documentation.

joanrue committed:
[_dev/L1Norm][BugFix] wrong output shape

joanrue committed:
[abc/operator][svdvals] return array in svdvals(), get item in lispchitz

joanrue committed:
[solver/pds] include option for ergodic estimate

joanrue committed:
[solver/pds][PD3O] add optimize_step_sizes

joanrue committed:
[solver/pds] refactor CB/DR/FB/PP // add DY/LV // add DiffMap support
Because Chambolle-Pock, Douglas-Rachford and Forward-Backward can be based either on Condat-Vu or PD3O, now they are defined as functions that return an object from the chosen base class. Add DavisYin and LorisVerhoeven algorithms. Also add minimal support composition of H w/ DiffMaps

joanrue committed:
[references] add ref to nonlinear Chambolle-Pock

matthieumeo committed:
[PDS] Remove ergodic variable and revert to old m_init architecture. More efficient PD3O iterations (one call only to adjoint and grad). PD3o's rho defaults to 1 to have the asymptotic rates of convergence guaranteed.

joanrue committed:
[PD3O/optimize_step_sizes] prevent effect of numerical error

joanrue committed:
[PD3O][DOC] remove ergodic remark

joanrue committed:
[PD3O][DOC] specify rho param. for convergence rate

joanrue committed:
[PD3O][BuxFix] add missing parameters

matthieumeo committed:
[PDS] Implement three strategies for setting the hyperparameters of CondatVu and PD3O (standard, large steps, overrelax) + updated documentation.

matthieumeo committed:
[PDS] Updates to the documentation of CP and DR + change in the default parameters of DR.

joanrue committed:
[solver/PDS] typos and small bugs

joanrue committed:
[solver/FB] instantiate it from Condat-Vu

joanrue committed:
[solver/PDS][DOC] add doc on tuning_strategy

joanrue committed:
[solver/DY][_set_step_sizes] overwrite method with DY conditions

joanrue committed:
[solver/LV][_set_step_sizes] overwrite method with LV conditions

joanrue committed:
[solver/PDS][add ADMM] algorithm 130 of [PSA]_

joanrue committed:
[solver/PDS] add notimplemented m_step method in base class

joanrue committed:
[operator/func] add QuadraticFunc

joanrue committed:
[refs/PDS] add references for ForwardBackward and ProximalPoint methods

joanrue committed:
[operator/arithmetic][BugFix] allow linop subclasses to be squeezed
I introduced this bug myself when I added the NonProxLike condition.

matthieumeo committed:
[operator] Remove the squeeze() fix (already fixed in a more robust way in v2-dev). Revert to NDArray as type annotation for the output of svdvals (think of the case k>1).

matthieumeo committed:
[func] Bug fixes and small improvements to QuadraticFunc.

matthieumeo committed:
[PDS] Small edits.

joanrue committed:
[PDS][_set_momentum_term] move children's method to parent class

joanrue committed:
[PDS][DOC] improve doc of LV and DY

joanrue committed:
[PDS/LV][REF] use base class' _set_step_size method

joanrue committed:
[PDS/ADMM] remove unused argument

joanrue committed:
[QuadraticFunc] change attributes to private

joanrue committed:
[PDS] resolve merging conflicts with v2-dev

joanrue committed:
[PDS] resolve merging conflicts with v2-dev (2)

matthieumeo committed:
[operator/func/base] Bug fixes in QuadraticFunc + implemented specific arithmetic for QuadraticFuncs

matthieumeo committed:
[opt/pds] Reordering of the various algorithms. Solution can now return the dual variable as well (for primal/dual algorithms). ADMM bugs fixed.

matthieumeo committed:
[opt/pds] Fixed merge issues.

matthieumeo committed:
[pds][init] Use kwargs to avoid cascading Solver.init() changes.

matthieumeo committed:
[pds] Enable automatic objective function tracking.

matthieumeo committed:
Merge branch 'v2-dev' into pds
